Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp

```diff
@@ -109,34 +204,99 @@ namespace AZ::DocumentPropertyEditor
+                                AZStd::make_shared<ReflectionAdapter>(m_reservedElementInstance, keyType);
                             Nodes::Adapter::QueryKey.InvokeOnDomNode(impl->m_adapter->GetContents(), &reflectionAdapter, path);
                             return; // key queried; don't store the actual entry until the DPE handles the QueryKey message
                         }
                     }
                 }
 
-                m_container->StoreElement(m_instance, dataAddress);
+                m_container->StoreElement(m_containerInstance, m_reservedElementInstance);
+                m_reservedElementInstance = nullptr;
 
                 auto containerNode = GetContainerNode(impl, path);
-                Nodes::PropertyEditor::AddNotify.InvokeOnDomNode(containerNode);
                 Nodes::PropertyEditor::ChangeNotify.InvokeOnDomNode(containerNode);
-
                 impl->m_adapter->NotifyResetDocument();
             }
 
             void OnRemoveElement(ReflectionAdapterReflectionImpl* impl, const AZ::Dom::Path& path)
             {
-                AZ_Assert(m_elementInstance != nullptr, "Attempted to remove an element without a defined element instance");
-                m_container->RemoveElement(m_instance, m_elementInstance, impl->m_serializeContext);
+                void* instanceToRemove = nullptr;
+                if (m_elementInstance == nullptr)
+                {
+                    instanceToRemove = m_containerInstance;
+                    auto associativeContainer = m_parentContainer->GetAssociativeContainerInterface();
+                    if (associativeContainer)
+                    {
+                        // Use the override instance if specified since it represents the
+                        // element pair instance that the IDataContainer API is expecting
+                        void* keyInstance = m_elementOverrideInstance ? m_elementOverrideInstance : m_containerInstance;
+                        const AZ::SerializeContext::ClassElement* containerClassElement =
+                            m_parentContainer->GetElement(m_parentContainer->GetDefaultElementNameCrc());
+                        instanceToRemove =
+                            associativeContainer->GetElementByKey(m_parentContainerInstance, containerClassElement, keyInstance);
+                    }
+                }
+                else
+                {
+                    instanceToRemove = m_elementInstance;
+                    auto associativeContainer = m_parentContainer->GetAssociativeContainerInterface();
+                    if (associativeContainer)
+                    {
+                        instanceToRemove = associativeContainer->GetElementByKey(m_parentContainerInstance, nullptr, m_elementInstance);
+                    }
+                }
+
+                m_parentContainer->RemoveElement(m_parentContainerInstance, instanceToRemove, impl->m_serializeContext);
```

Comment:
`instanceToRemove` needs to be check that it is not nullptr before RemoveElement is called.
GetElementByKey can return nullptr

Reply:
Done.